# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [64]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [68]:
path = os.getcwd()
ski_data = pd.read_csv(path + '/data/step3_output.csv' , index_col = 0)
print(ski_data.head())

                             Name       state  summit_elev  vertical_drop  \
2                Hilltop Ski Area      Alaska         2090            294   
4             Sunrise Park Resort     Arizona        11100           1800   
5   Yosemite Ski & Snowboard Area  California         7800            600   
10               Donner Ski Ranch  California         8012            750   
12                  June Mountain  California        10090           2590   

    trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
2       0        0.0          0          0     0       1  ...   
4       0        0.0          0          1     2       3  ...   
5       0        0.0          0          0     0       1  ...   
10      0        0.0          0          0     0       1  ...   
12      0        0.0          0          2     0       0  ...   

    SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
2                30.0       30.000000        150.000000       36.0   
4     

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [69]:
ski_data_d = pd.concat([ski_data.drop('state', axis = 1), pd.get_dummies(ski_data['state'])], axis = 1)
print(ski_data_d.tail())

                                  Name  summit_elev  vertical_drop  trams  \
323                      Hogadon Basin         8000            640      0   
326          Sleeping Giant Ski Resort         7428            810      0   
327                   Snow King Resort         7808           1571      0   
328  Snowy Range Ski & Recreation Area         9663            990      0   
329                White Pine Ski Area         9500           1100      0   

     fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
323        0.0          0          0     0       0       1  ...             0   
326        0.0          0          0     0       1       1  ...             0   
327        0.0          0          0     1       1       1  ...             0   
328        0.0          0          0     0       1       3  ...             0   
329        0.0          0          0     0       2       0  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virgi

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [86]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = ski_data_d.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = ski_data_d['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [87]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = np.ravel(y)

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [88]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [89]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [0]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [90]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

-4.129358353873719e+22
737280636500.4358


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [91]:
print(model.intercept_)

185423872707.78036


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [92]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coef = coef.sort_values('Coefficient', ascending = False)
print(coef.head(10))

               Coefficient
total_chairs  2.029647e+13
fastEight     1.608656e+13
surface       1.159482e+13
double        1.122069e+13
fastSixes     1.065419e+13
triple        9.575886e+12
quad          5.790353e+12
trams         4.921148e+12
fastQuads     4.604648e+12
New York      3.277513e+12


**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [93]:
X = ski_data.drop(['Name', 'state', 'AdultWeekend'], axis = 1)

y = ski_data['AdultWeekend']

scaler = preprocessing.StandardScaler().fit(X)

X_scaled=scaler.transform(X) 

y = np.ravel(y)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [94]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

print(model.intercept_)

coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coef = coef.sort_values('Coefficient', ascending = False)
print(coef.head(10))

0.5559565165902214
6.85299068624076
56.241296374899306
                  Coefficient
AdultWeekday        10.986869
summit_elev          3.408969
clusters             2.933922
daysOpenLastYear     2.063798
averageSnowfall      1.916504
Runs                 1.563800
vertical_drop        1.417618
triple               1.327128
surface              1.229705
NightSkiing_ac       0.804283


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [95]:
X = ski_data.drop(['Name', 'state', 'AdultWeekend', 'summit_elev'], axis = 1)

y = ski_data['AdultWeekend']

scaler = preprocessing.StandardScaler().fit(X)

X_scaled=scaler.transform(X) 

y = np.ravel(y)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [96]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

print(model.intercept_)

coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coef = coef.sort_values('Coefficient', ascending = False)
print(coef.head(10))

0.5413080297878097
7.275361831392762
56.169716762833055
                   Coefficient
AdultWeekday         10.718374
averageSnowfall       2.856805
clusters              2.234782
vertical_drop         1.905925
SkiableTerrain_ac     1.761634
daysOpenLastYear      1.459799
triple                1.229914
Runs                  1.099987
NightSkiing_ac        1.028189
surface               0.901449


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -4.129358353873719e+22 |737280636500.4358  |-|
| Model 2. | 0.5559565165902214|6.85299068624076 |'state'|
| Model 3. | 0.5413080297878097 |7.275361831392762 |'state','summit_elev','base_elev'|

Model Selection: Model 3 is the best choice for moving forward with the Big Mountain Ski Resort project. Its performance is comparable to that of Model 2 (only slightly worse), and it provides more actionable insights for resort management.